src/network/stream.rs

```diff
@@ -0,0 +1,50 @@
+use libp2p::{core::ConnectedPoint, PeerId};
+use tokio::sync::{mpsc, Mutex};
+use tokio_stream::wrappers::ReceiverStream;
+
+#[derive(Debug, Clone)]
+pub enum Event {
+	ConnectionEstablished {
+		peer_id: PeerId,
+		endpoint: ConnectedPoint,
+	},
+}
+struct EventStream {
+	senders: Vec<mpsc::Sender<Event>>,
+}
+
+impl EventStream {
+	fn stream(&mut self) -> ReceiverStream<Event> {
+		let (tx, rx) = mpsc::channel(1000);
+		self.senders.push(tx);
+		ReceiverStream::new(rx)
+	}
+
+	fn notify(&mut self, event: Event) {
+		self.senders.retain(|tx| tx.try_send(event.clone()).is_ok());
+	}
+}
+
+pub struct NetworkStream {
+	stream: Mutex<EventStream>,
+}
+
+impl NetworkStream {
+	pub fn new() -> Self {
+		Self {
+			stream: Mutex::new(EventStream {
```

Comment:
I think (hope) we can derive default on EventStream maybe even on NetworkStream and have only Self::default() or something like that.